### ISSUE TITLE
masking on tensor.data?

### ISSUE THREAD

**Issue by chihyaoma** [jadore801120/attention-is-all-you-need-pytorch#16]
Hi, 

I noticed that you were masking out the padded tensor by assigning value to `tensor.data`.

```
attn.data.masked_fill_(attn_mask, -float('inf'))
```

Is this correct? 

Based on the discussion [here](https://discuss.pytorch.org/t/partially-reset-a-variable-in-place-vs-new-variable/936/2), shouldn't we assign values to tensor itself, instead of tensor.data? In this way, the history of the gradient can be tracked.

**Comment by ZiJianZhao** [comment]
Maybe you are right. However, even if I change that like what you say. I can't run into meaningful results. The generated text of beam search is wrong. And the accuracy is not a reliable indicator. So do you have right generated text?

**Comment by chihyaoma** [comment]
I didn't actually run the code. I was just curious about the implementation and notice this potential issue. 

Just curious. When assigning values to the Variable itself, have you tried assigned it to `0`, instead of `-float('inf')`?

**Comment by ZiJianZhao** [comment]
It is wrong to assign `0` because a softmax layer is following. If you assign that position to` 0`, then you would get `exp(0)=1` in the masked position. But it should be `float('-inf')` which leads `exp(flaot('-inf')=0`.

**Comment by jadore801120** [comment]
Hello @chihyaoma, thanks again for the notice. with the newest api, I have avoided to manipluate on the data itself. Please take a look.
Hello @ZiJianZhao , thanks for explanation!
Thanks,
Yu-Hsiang
